app/controllers/api/storage.php

```diff
@@ -245,6 +246,155 @@
+            throw new Exception('Invalid file', 403);
+        }
+
+        // Save to storage
+        $size = $device->getFileSize($file['tmp_name']);
+        $path = $device->getPath(\uniqid().'.'.\pathinfo($file['name'], PATHINFO_EXTENSION));
+        $path = $bucket->getId() . '/' . $path;
+        
+        if (!$device->upload($file['tmp_name'], $path)) { // TODO deprecate 'upload' and replace with 'move'
+            throw new Exception('Failed moving file', 500);
+        }
+
+        $mimeType = $device->getFileMimeType($path); // Get mime-type before compression and encryption
+
+        if (App::getEnv('_APP_STORAGE_ANTIVIRUS') === 'enabled' && $bucket->getAttribute('antiVirus', true) && $size <= APP_LIMIT_ANTIVIRUS) {
+            $antiVirus = new Network(App::getEnv('_APP_STORAGE_ANTIVIRUS_HOST', 'clamav'),
+                (int) App::getEnv('_APP_STORAGE_ANTIVIRUS_PORT', 3310));
+
+            if (!$antiVirus->fileScan($path)) {
+                $device->delete($path);
+                throw new Exception('Invalid file', 403);
+            }
+        }
+
+        // Compression
+        $data = $device->read($path);
+        if($size <= APP_LIMIT_COMPRESSION) {
+            $compressor = new GZIP();
+            $data = $compressor->compress($data);
+        }
+        
+        if($bucket->getAttribute('encryption', true) && $size <= APP_LIMIT_ENCRYPTION) {
+            $key = App::getEnv('_APP_OPENSSL_KEY_V1');
+            $iv = OpenSSL::randomPseudoBytes(OpenSSL::cipherIVLength(OpenSSL::CIPHER_AES_128_GCM));
+            $data = OpenSSL::encrypt($data, OpenSSL::CIPHER_AES_128_GCM, $key, 0, $iv, $tag);
+        }
+
+        if (!$device->write($path, $data, $mimeType)) {
+            throw new Exception('Failed to save file', 500);
+        }
+
+        $sizeActual = $device->getFileSize($path);
+        
+        $data = [
+            '$read' => (is_null($read) && !$user->isEmpty()) ? ['user:'.$user->getId()] : $read ?? [], // By default set read permissions for user
+            '$write' => (is_null($write) && !$user->isEmpty()) ? ['user:'.$user->getId()] : $write ?? [], // By default set write permissions for user
+            'dateCreated' => \time(),
+            'bucketId' => $bucket->getId(),
+            'name' => $file['name'],
+            'path' => $path,
+            'signature' => $device->getFileHash($path),
+            'mimeType' => $mimeType,
+            'sizeOriginal' => $size,
+            'sizeActual' => $sizeActual,
+            'algorithm' => empty($compressor) ? '' : $compressor->getName(),
+            'comment' => '',
+        ];
+
+        if($bucket->getAttribute('encryption', true)) {
+            $data['openSSLVersion'] = '1';
+            $data['openSSLCipher'] = OpenSSL::CIPHER_AES_128_GCM;
+            $data['openSSLTag'] = \bin2hex($tag);
+            $data['openSSLIV'] = \bin2hex($iv);
+        }
+
+        $file = $dbForInternal->createDocument('files', new Document($data));
```

Comment:
What do we do when a file is uploaded, but couldn't be saved to the database?

Reply:
Let's tackle it later :D

Reply:
Maybe we can create some sort of a periodic cleanup?

Reply:
Anyway, we can tackle it later :)